### ISSUE TITLE
Use PDO and parameterised queries

### ISSUE THREAD

**Issue by Simon-Davies** [joomla/joomla-cms#16094]
In the light of the recent SQL Injection vulnerability in Joomla 3.7 would it not be a good idea to rewrite Joomla's code to use PDO and parameterised queries?

I know this would be a lot of work but I'm willing to help.

**Comment by mbabker** [comment]
It's not possible during 3.x, `ext/mysql` doesn't support prepared statements and it's not worth it to emulate this, especially as we're dropping it in 4.0.

As for the rest of the package, the bulk of the work is already done but the implementations are more reliant on each driver's configuration so right now I'm not sure even with all the work ported over we could write SQL and handle parameterized queries all that well without having an extra layer of string processing in our query processing code.

This is the stuff that needs to be ported over:

- https://github.com/joomla-framework/database/commit/66035e44f523bb0259aadb0e544fc8cf63e686d2 - Adds bound variables to the PDO MySQL query object (note the abstract PDO driver already supports prepared statements)
- https://github.com/joomla-framework/database/pull/40 - PDO implementation of PostgreSQL, usable in parallel with the native `ext/pgsql` support
- https://github.com/joomla-framework/database/pull/38 - MySQLi prepared statement support
- https://github.com/joomla-framework/database/pull/39 - PostgreSQL prepared statement support
- https://github.com/joomla-framework/database/pull/78 - SQL Server prepared statement support

**Comment by SniperSister** [comment]
Count me in here too, would love to see that happen!

**Comment by mbabker** [comment]
#16402 would make use of the Framework database package in 4.0.  Includes making prepared statement support available and PDO PostgreSQL connections.

**Comment by brianteeman** [comment]
closed as this work is in progress as mentioned above.
